src/server/client.ts

```diff
@@ -0,0 +1,505 @@
+/// <reference path="session.ts" />
+
+module ts.server {
+
+    export interface SessionClientHost extends LanguageServiceHost {
+        writeMessage(message: string): void;
+    }
+
+    interface CompletionEntry extends CompletionInfo {
+        fileName: string;
+        position: number;
+    }
+
+    interface RenameEntry extends RenameInfo {
+        fileName: string;
+        position: number;
+        locations: RenameLocation[];
+        findInStrings: boolean;
+        findInComments: boolean;
+    }
+
+    export class SessionClient implements LanguageService {
+        private sequence: number = 0;
+        private fileMapping: ts.Map<string> = {};
+        private lineMaps: ts.Map<number[]> = {};
+        private messages: string[] = [];
+        private lastCompletionEntry: CompletionEntry;
+        private lastRenameEntry: RenameEntry;
+        
+        constructor(private host: SessionClientHost) {
+        }
+
+        public onMessage(message: string): void { 
+            this.messages.push(message);
+        }
+
+        private writeMessage(message: string): void { 
+            this.host.writeMessage(message);
+        }
+
+        private getLineMap(fileName: string): number[] { 
+            var lineMap = ts.lookUp(this.lineMaps, fileName);
+            if (!lineMap) {
+                var scriptSnapshot = this.host.getScriptSnapshot(fileName);
+                lineMap  = this.lineMaps[fileName] = ts.computeLineStarts(scriptSnapshot.getText(0, scriptSnapshot.getLength()));
+            }
+            return lineMap;
+        }
+
+        private lineColToPosition(fileName: string, lineCol: ts.server.protocol.LineCol): number {

```

Comment:
LineCol is a bad name. It implies 'Column'.  but Column is, intrinsically, a host concept.  For example, if you have a tab, then Column might be 8 with one set of settings, but could also be 4 with different settings.  LineCharacter is likely more what we want.